Either coders throw ClassCastException in Scio 0.8.0

On Scio 0.8.0:

Open SBT test console:

```
sbt
> project scio-core
> test:console
```

Paste the following into the console:

```
import com.spotify.scio.coders._
import com.spotify.scio._
case class Example(stringT: Either[Array[Byte], String], longT: Either[Array[Byte], Long])
val sc = ScioContext.forTest()
val c = CoderMaterializer.beam(sc, implicitly[Coder[Example]])
c.encode(Example(Right("str"), Right(0L)), System.out)
```

Expected:
Coder should throw no exceptions and successfully encode an Example case class containing two Eithers of different types

Actual: 
The following exception is thrown: 

```
java.lang.RuntimeException: Exception while trying to `encode` an instance of Example:  Can't encode field longT value Right(0)
  at com.spotify.scio.coders.RecordCoder.onErrorMsg(Coder.scala:272)
  at com.spotify.scio.coders.RecordCoder.encode(Coder.scala:284)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply$mcV$sp(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:216)
  at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:226)
  ... 47 elided
Caused by: java.lang.RuntimeException: Exception while trying to `encode` an instance of scala.util.Right:  Can't encode field value value 0
  at com.spotify.scio.coders.RecordCoder.onErrorMsg(Coder.scala:272)
  at com.spotify.scio.coders.RecordCoder.encode(Coder.scala:284)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply$mcV$sp(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:216)
  at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:226)
  at com.spotify.scio.coders.DisjunctionCoder.encode(Coder.scala:126)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply$mcV$sp(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:216)
  at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:226)
  at com.spotify.scio.coders.RefCoder$$anonfun$encode$1.apply$mcV$sp(Coder.scala:184)
  at com.spotify.scio.coders.RefCoder$$anonfun$encode$1.apply(Coder.scala:184)
  at com.spotify.scio.coders.RefCoder$$anonfun$encode$1.apply(Coder.scala:184)
  at com.spotify.scio.coders.RefCoder.check(Coder.scala:180)
  at com.spotify.scio.coders.RefCoder.encode(Coder.scala:184)
  at com.spotify.scio.coders.RecordCoder$$anonfun$encode$4.apply$mcV$sp(Coder.scala:284)
  at com.spotify.scio.coders.RecordCoder$$anonfun$encode$4.apply(Coder.scala:284)
  at com.spotify.scio.coders.RecordCoder$$anonfun$encode$4.apply(Coder.scala:284)
  at com.spotify.scio.coders.RecordCoder.onErrorMsg(Coder.scala:269)
  at com.spotify.scio.coders.RecordCoder.encode(Coder.scala:284)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply$mcV$sp(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:216)
  at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:226)
Caused by: java.lang.ClassCastException: java.lang.Long cannot be cast to java.lang.String
  at org.apache.beam.sdk.coders.StringUtf8Coder.encode(StringUtf8Coder.java:37)
  at com.spotify.scio.coders.RecordCoder$$anonfun$encode$4.apply$mcV$sp(Coder.scala:284)
  at com.spotify.scio.coders.RecordCoder$$anonfun$encode$4.apply(Coder.scala:284)
  at com.spotify.scio.coders.RecordCoder$$anonfun$encode$4.apply(Coder.scala:284)
  at com.spotify.scio.coders.RecordCoder.onErrorMsg(Coder.scala:269)
  at com.spotify.scio.coders.RecordCoder.encode(Coder.scala:284)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply$mcV$sp(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:216)
  at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:226)
  at com.spotify.scio.coders.DisjunctionCoder.encode(Coder.scala:126)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply$mcV$sp(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:216)
  at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:226)
  at com.spotify.scio.coders.RefCoder$$anonfun$encode$1.apply$mcV$sp(Coder.scala:184)
  at com.spotify.scio.coders.RefCoder$$anonfun$encode$1.apply(Coder.scala:184)
  at com.spotify.scio.coders.RefCoder$$anonfun$encode$1.apply(Coder.scala:184)
  at com.spotify.scio.coders.RefCoder.check(Coder.scala:180)
  at com.spotify.scio.coders.RefCoder.encode(Coder.scala:184)
  at com.spotify.scio.coders.RecordCoder$$anonfun$encode$4.apply$mcV$sp(Coder.scala:284)
  at com.spotify.scio.coders.RecordCoder$$anonfun$encode$4.apply(Coder.scala:284)
  at com.spotify.scio.coders.RecordCoder$$anonfun$encode$4.apply(Coder.scala:284)
  at com.spotify.scio.coders.RecordCoder.onErrorMsg(Coder.scala:269)
  at com.spotify.scio.coders.RecordCoder.encode(Coder.scala:284)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply$mcV$sp(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$2.apply(Coder.scala:226)
  at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:216)
  at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:226)
```

--- a/scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala
@@ -171,31 +171,26 @@ final private case class DisjunctionCoder[T, Id](
     }
 }
 
-final private[scio] case class RefCoder[T](typeName: String, var coder: BCoder[T])
+final private[scio] case class LazyCoder[T](
+  typeName: String,
+  o: CoderMaterializer.CoderOptions
+)(coder: Coder[T])
     extends BCoder[T] {
-  def setImpl(_c: BCoder[T]): Unit = coder = _c
 
-  private def check[A](t: => A): A = {
-    require(coder != null, s"Coder implementation should not be null in ${this}")
-    t
-  }
+  private lazy val bcoder = CoderMaterializer.beamImpl[T](o, coder)
 
-  def decode(inStream: InputStream): T = check { coder.decode(inStream) }
-  def encode(value: T, outStream: OutputStream): Unit = check { coder.encode(value, outStream) }
-  def getCoderArguments(): java.util.List[_ <: BCoder[_]] = check { coder.getCoderArguments() }
-  def verifyDeterministic(): Unit = check { coder.verifyDeterministic() }
+  def decode(inStream: InputStream): T = bcoder.decode(inStream)
+  def encode(value: T, outStream: OutputStream): Unit = bcoder.encode(value, outStream)
+  def getCoderArguments(): java.util.List[_ <: BCoder[_]] = bcoder.getCoderArguments()
+  def verifyDeterministic(): Unit = bcoder.verifyDeterministic()
 
-  override def consistentWithEquals(): Boolean = check { coder.consistentWithEquals() }
-  override def structuralValue(value: T): AnyRef = check {
+  override def consistentWithEquals(): Boolean = bcoder.consistentWithEquals()
+  override def structuralValue(value: T): AnyRef =
     if (consistentWithEquals()) {
       value.asInstanceOf[AnyRef]
     } else {
-      coder.structuralValue(value)
+      bcoder.structuralValue(value)
     }
-  }
-
-  override def toString(): String =
-    s"RefCoder($typeName, ${if (coder == null) "null" else "<coder ref>"})"
 }
 
 // XXX: Workaround a NPE deep down the stack in Beam

--- a/scio-core/src/main/scala/com/spotify/scio/coders/CoderMaterializer.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/coders/CoderMaterializer.scala
@@ -19,10 +19,17 @@ package com.spotify.scio.coders
 
 import org.apache.beam.sdk.coders.{CoderRegistry, KvCoder, NullableCoder, Coder => BCoder}
 import org.apache.beam.sdk.options.{PipelineOptions, PipelineOptionsFactory}
-import scala.collection.concurrent.TrieMap
 
 object CoderMaterializer {
   import com.spotify.scio.ScioContext
+
+  private[scio] case class CoderOptions(nullableCoders: Boolean, kryo: KryoOptions)
+  private[scio] object CoderOptions {
+    def apply(o: PipelineOptions) = {
+      val nullableCoder = o.as(classOf[com.spotify.scio.options.ScioOptions]).getNullableCoders()
+      new CoderOptions(nullableCoder, KryoOptions(o))
+    }
+  }
 
   final def beam[T](sc: ScioContext, c: Coder[T]): BCoder[T] =
     beam(sc.pipeline.getCoderRegistry, sc.options, c)
@@ -33,23 +40,19 @@ object CoderMaterializer {
     o: PipelineOptions = PipelineOptionsFactory.create()
   ): BCoder[T] = beam(r, o, coder)
 
-  @inline private def nullCoder[T](o: PipelineOptions, c: BCoder[T]) = {
-    val nullableCoder = o.as(classOf[com.spotify.scio.options.ScioOptions]).getNullableCoders()
-    if (nullableCoder) NullableCoder.of(c)
+  @inline private def nullCoder[T](o: CoderOptions, c: BCoder[T]) =
+    if (o.nullableCoders) NullableCoder.of(c)
     else c
-  }
 
   final def beam[T](
     r: CoderRegistry,
     o: PipelineOptions,
     coder: Coder[T]
-  ): BCoder[T] = beamImpl(r, o, coder, TrieMap.empty)
+  ): BCoder[T] = beamImpl(CoderOptions(o), coder)
 
-  final private def beamImpl[T](
-    r: CoderRegistry,
-    o: PipelineOptions,
-    coder: Coder[T],
-    refs: TrieMap[String, RefCoder[_]]
+  final private[scio] def beamImpl[T](
+    o: CoderOptions,
+    coder: Coder[T]
   ): BCoder[T] =
     coder match {
       // #1734: do not wrap native beam coders
@@ -58,16 +61,16 @@ object CoderMaterializer {
       case Beam(c) =>
         WrappedBCoder.create(nullCoder(o, c))
       case Fallback(_) =>
-        val kryoCoder = new KryoAtomicCoder[T](KryoOptions(o))
+        val kryoCoder = new KryoAtomicCoder[T](o.kryo)
         WrappedBCoder.create(nullCoder(o, kryoCoder))
       case Transform(c, f) =>
-        val u = f(beamImpl(r, o, c, refs))
-        WrappedBCoder.create(beamImpl(r, o, u, refs))
+        val u = f(beamImpl(o, c))
+        WrappedBCoder.create(beamImpl(o, u))
       case Record(typeName, coders, construct, destruct) =>
         WrappedBCoder.create(
           new RecordCoder(
             typeName,
-            coders.map(c => c._1 -> nullCoder(o, beamImpl(r, o, c._2, refs))),
+            coders.map(c => c._1 -> nullCoder(o, beamImpl(o, c._2))),
             construct,
             destruct
           )
@@ -77,25 +80,15 @@ object CoderMaterializer {
           // `.map(identity) is really needed to make Map serializable.
           DisjunctionCoder(
             typeName,
-            beamImpl(r, o, idCoder, refs),
+            beamImpl(o, idCoder),
             id,
-            coders.mapValues(u => beamImpl(r, o, u, refs)).map(identity)
+            coders.mapValues(u => beamImpl(o, u)).map(identity)
           )
         )
       case KVCoder(koder, voder) =>
-        WrappedBCoder.create(KvCoder.of(beamImpl(r, o, koder, refs), beamImpl(r, o, voder, refs)))
+        WrappedBCoder.create(KvCoder.of(beamImpl(o, koder), beamImpl(o, voder)))
       case Ref(t, c) =>
-        refs
-          .get(t)
-          .getOrElse {
-            // #2269: _REALLY_ support recursive ADTs
-            val bc = RefCoder[T](t, null) // placeholder
-            refs.put(t, bc)
-            val actualImpl = beamImpl(r, o, c, refs)
-            bc.setImpl(actualImpl)
-            bc
-          }
-          .asInstanceOf[BCoder[T]]
+        LazyCoder[T](t, o)(c)
     }
 
   def kvCoder[K, V](ctx: ScioContext)(implicit k: Coder[K], v: Coder[V]): KvCoder[K, V] =

--- a/scio-core/src/main/scala/com/spotify/scio/coders/DerivedCoder.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/coders/DerivedCoder.scala
@@ -36,38 +36,41 @@ private object Derived extends Serializable {
     typeName: TypeName,
     ps: Seq[Param[Coder, T]],
     rawConstruct: Seq[Any] => T
-  ): Coder[T] = {
-    val cs = new Array[(String, Coder[Any])](ps.length)
-    var i = 0
-    while (i < ps.length) {
-      val p = ps(i)
-      cs.update(i, (p.label, p.typeclass.asInstanceOf[Coder[Any]]))
-      i = i + 1
-    }
-
-    val materializationStack = CoderStackTrace.prepare
-
-    @inline def destruct(v: T): Array[Any] = {
-      val arr = new Array[Any](ps.length)
-      var i = 0
-      while (i < ps.length) {
-        val p = ps(i)
-        catching(s"Error while dereferencing parameter ${p.label} in $v", materializationStack) {
-          arr.update(i, p.dereference(v))
+  ): Coder[T] =
+    Ref(
+      typeName.full, {
+        val cs = new Array[(String, Coder[Any])](ps.length)
+        var i = 0
+        while (i < ps.length) {
+          val p = ps(i)
+          cs.update(i, (p.label, p.typeclass.asInstanceOf[Coder[Any]]))
           i = i + 1
         }
+
+        val materializationStack = CoderStackTrace.prepare
+
+        @inline def destruct(v: T): Array[Any] = {
+          val arr = new Array[Any](ps.length)
+          var i = 0
+          while (i < ps.length) {
+            val p = ps(i)
+            catching(s"Error while dereferencing parameter ${p.label} in $v", materializationStack) {
+              arr.update(i, p.dereference(v))
+              i = i + 1
+            }
+          }
+          arr
+        }
+
+        val constructor: Seq[Any] => T =
+          ps =>
+            catching(s"Error while constructing object from parameters $ps", materializationStack)(
+              rawConstruct(ps)
+            )
+
+        Coder.record[T](typeName.full, cs, constructor, destruct)
       }
-      arr
-    }
-
-    val constructor: Seq[Any] => T =
-      ps =>
-        catching(s"Error while constructing object from parameters $ps", materializationStack)(
-          rawConstruct(ps)
-        )
-
-    Coder.record[T](typeName.full, cs, constructor, destruct)
-  }
+    )
 }
 
 trait LowPriorityCoderDerivation {

--- a/scio-macros/src/main/scala/com/spotify/scio/MagnoliaMacros.scala
+++ b/scio-macros/src/main/scala/com/spotify/scio/MagnoliaMacros.scala
@@ -36,7 +36,7 @@ private[scio] object MagnoliaMacros {
     val magnoliaTree = magnolia.Magnolia.gen[T](c)
 
     // Remove annotations from magnolia since they are
-    // not serialiazable and we don't use them anyway
+    // not serializable and we don't use them anyway
 
     val removeAnnotations = new Transformer {
       override def transform(tree: Tree): c.universe.Tree =

--- a/scio-test/src/test/scala/com/spotify/scio/coders/CoderTest.scala
+++ b/scio-test/src/test/scala/com/spotify/scio/coders/CoderTest.scala
@@ -489,6 +489,11 @@ final class CoderTest extends AnyFlatSpec with Matchers {
     val c = CoderMaterializer.beam(sc, implicitly[Coder[Example]])
     c.encode(Example(Right("str"), Right(0L)), System.out)
   }
+
+  it should "#2467 support derivation of directly recursive types" in {
+    import RecursiveCase._
+    Recursive(1, Option(Recursive(2, None))) coderShould roundtrip()
+  }
 }
 
 object RecursiveCase {
@@ -499,4 +504,6 @@ object RecursiveCase {
   case class RecordType(fields: List[SampleField]) extends SampleFieldType
 
   implicit val coderSampleFieldType = Coder.gen[SampleField]
+
+  case class Recursive(a: Int, rec: Option[Recursive] = None)
 }

--- a/scio-test/src/test/scala/com/spotify/scio/coders/CoderTest.scala
+++ b/scio-test/src/test/scala/com/spotify/scio/coders/CoderTest.scala
@@ -482,6 +482,13 @@ final class CoderTest extends AnyFlatSpec with Matchers {
       )
     ) coderShould roundtrip()
   }
+
+  it should "#2595: work with parameterized types" in {
+    case class Example(stringT: Either[Array[Byte], String], longT: Either[Array[Byte], Long])
+    val sc = com.spotify.scio.ScioContext.forTest()
+    val c = CoderMaterializer.beam(sc, implicitly[Coder[Example]])
+    c.encode(Example(Right("str"), Right(0L)), System.out)
+  }
 }
 
 object RecursiveCase {

